src/Identity/Controllers/AccountsController.cs

```diff
@@ -217,11 +274,24 @@ public async Task<PreloginResponseModel> PostPrelogin([FromBody] PreloginRequest
         var kdfInformation = await _userRepository.GetKdfInformationByEmailAsync(model.Email);
         if (kdfInformation == null)
         {
-            kdfInformation = new UserKdfInformation
+            if (_defaultKdfHmacKey == null)
             {
-                Kdf = KdfType.PBKDF2_SHA256,
-                KdfIterations = AuthConstants.PBKDF2_ITERATIONS.Default,
-            };
+                kdfInformation = _defaultKdfResults[0];
+            }
+            else
+            {
+                // Compute the HMAC hash of the email
+                var hmacMessage = Encoding.UTF8.GetBytes(model.Email.Trim().ToLowerInvariant());
+                using var hmac = new System.Security.Cryptography.HMACSHA256(_defaultKdfHmacKey);
+                var hmacHash = hmac.ComputeHash(hmacMessage);
+                // Convert the hash to a number
+                var hashHex = BitConverter.ToString(hmacHash).Replace("-", string.Empty).ToLowerInvariant();
+                var hashFirst8Bytes = hashHex.Substring(0, 16);
+                var hashNumber = long.Parse(hashFirst8Bytes, System.Globalization.NumberStyles.HexNumber);
+                // Find the default KDF value for this hash number
+                var hashIndex = (int)(Math.Abs(hashNumber) % _defaultKdfResults.Count);
+                kdfInformation = _defaultKdfResults[hashIndex];
```

Comment:
I'd recommend using `XxHash32` for this instead of `HMACSHA256`, not only is is much faster:

```
| Method            | Mean        |
|------------------ |------------:|
| HMACSHA256        | 1,602.88 ns |
| XxHash32          |    39.02 ns |

Tested with the email: " John.smith@gmail.com"
```

but since we won't be truncating any of it, it's much more dispersion tested. 

Here is the code for it:

```C#
var hmacMessage = Encoding.UTF8.GetBytes(_cases[EmailCase].Trim().ToLowerInvariant());
Span<byte> hash = stackalloc byte[sizeof(uint)];
XxHash32.Hash(hmacMessage, hash, _seed /* this replaces `_defaultKdfHmaxKey` */);
// Convert the hash to a number
var hashNumber = BitConverter.ToUInt32(hash);
// Find the default KDF value for this hash number
var hashIndex = (int)(hashNumber % _defaultKdfResults.Count);
return _defaultKdfResults[hashIndex];
```

Since I use a `stackalloc` this would have to move into a non-async method and then just called from the controller action.

Reply:
Nice. I have never used XxHash, but I am not sure the perf savings of HMAC vs XxHash are really substantial in the grand scheme of things here.